File: java-learning-path/practise/LeapYear.java
import java.util.Scanner;

/**
 * Exercise 01: Leap Year Checker
 * 
 * ALGORITHM:
 * 1. If a year is divisible by 4, it MIGHT be a leap year.
 * 2. If it is also divisible by 100, it is NOT a leap year UNLESS:
 * 3. It is also divisible by 400.
 */
public class LeapYear {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter a year: ");
        
        // Handle potential non-integer input gracefully
        if (!scanner.hasNextInt()) {
            System.out.println("Invalid input. Please enter a numeric year.");
            return;
        }
        
        int year = scanner.nextInt();

        boolean isLeap = false;
        
        // Step 1: Check divisibility by 4
        if (year % 4 == 0) {
            // Step 2: Century years (ending in 00) must be divisible by 400
            if (year % 100 == 0) {
                isLeap = (year % 400 == 0); // Divisible by 400 -> true, else -> false
            } else {
                isLeap = true; // Divisible by 4 but not by 100
            }
        }

        // Output the result
        if (isLeap) {
            System.out.println(year + " is a leap year.");
        } else {
            System.out.println(year + " is not a leap year.");
        }
        
        scanner.close();
    }
}

File: python-learning-path/practise/leap_year.py
"""
Exercise 01: Leap Year Checker

ALGORITHM:
1. Divide by 4: If yes, it's a leap year candidate.
2. Divide by 100: If yes, it's NOT a leap year UNLESS:
3. Divide by 400: If yes, it's definitely a leap year.
"""

def is_leap_year(year):
    # Pythonic logic using boolean operators
    # A year is a leap year if (divisible by 4 AND not by 100) OR (divisible by 400)
    return (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0)

def main():
    try:
        # Get input and convert to integer
        user_input = input("Enter a year: ").strip()
        if not user_input:
            print("Please enter a value.")
            return
            
        year = int(user_input)
        
        if is_leap_year(year):
            print(f"✅ {year} is a leap year.")
        else:
            print(f"❌ {year} is NOT a leap year.")
            
    except ValueError:
        print("⚠️ Error: Please enter a valid numeric year.")

if __name__ == "__main__":
    main()
